shared-module/usb_cdc/Serial.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 Dan Halbert for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "lib/utils/interrupt_char.h"
+#include "shared-module/usb_cdc/Serial.h"
+#include "supervisor/shared/tick.h"
+
+#include "tusb.h"
+
+size_t common_hal_usb_cdc_serial_read(usb_cdc_serial_obj_t *self, uint8_t *data, size_t len, int *errcode) {
+    if (self->timeout < 0.0f) {
+        while (tud_cdc_n_available(self->idx) < len) {
```

Comment:
Shouldn't we just return the bytes currently available instead of waiting for all of them?

Reply:
There are three cases:
- `timeout` < 0 corresponds to `timeout = None` on the Python side. A timeout of `None` in pyserial waits _indefinitely_ for the requested number of bytes. (But I should fix this so that we don't wait for more bytes than the buffer will ever hold - that's a bug.)
- `timeout` == 0 returns whatever bytes are available, and returns immediately.
- `timeout` > 0 waits for up to `timeout` seconds to try to get all the bytes requested. If the timeout expires, then return whatever bytes we have.

Does this make sense or am I misunderstanding your question?

Reply:
> (But I should fix this so that we don't wait for more bytes than the buffer will ever hold - that's a bug.)

And I should fill the passed buffer partially and then wait for more. Same for the `timeout > 0` case.

Reply:
Yup, makes sense. I usually implement 1 by passing 3 MAXINT.